numpy/core/numeric.py

```diff
@@ -2315,15 +2317,26 @@ def array_equal(a1, a2):
     False
     >>> np.array_equal([1, 2], [1, 4])
     False
-
+    >>> a = np.array([1, np.nan])
+    >>> np.array_equal(a, a)
+    False
+    >>> np.array_equal(a, a, equal_nan=True)
+    True
     """
     try:
         a1, a2 = asarray(a1), asarray(a2)
     except Exception:
         return False
     if a1.shape != a2.shape:
         return False
-    return bool(asarray(a1 == a2).all())
+    if not equal_nan:
+        return bool(asarray(a1 == a2).all())
+    # Handling NaN values if equal_nan is True
+    a1nan, a2nan = isnan(a1), isnan(a2)
+    # NaN's occur at different locations
+    if not all(a1nan == a2nan):
+        return False
+    return bool(asarray(a1[~a1nan] == a2[~a2nan]).all())
```

Comment:
Hehe, tricky. Maybe you can add a comment that the boolean indexing works because `a1` and `a2` must have the same shape (in fact, you could just use `a1nan` for both, since they are equivalent to make that more obvious).

Reply:
:+1: I agree that it might improve clarity.